File: ios/Fishers/ViewModels/ChatStore.swift
import Foundation
import SwiftUI

/// Chat state: the thread list, the open thread's messages, and the assistant's
/// proposals. Proposals are applied here so the thread refreshes with the note
/// the server posts.
@MainActor
final class ChatStore: ObservableObject {
    @Published var conversations: [ConversationSummary] = []
    @Published var messages: [ChatMessage] = []
    @Published var proposals: [AgentProposal] = []
    @Published var isLoading = false
    @Published var isSending = false
    @Published var isThinking = false
    @Published var errorMessage: String?
    @Published var agentSummary: String?

    private var openConversationId: UUID?

    var pendingProposals: [AgentProposal] { proposals.filter(\.isPending) }

    func loadConversations() async {
        isLoading = true
        defer { isLoading = false }
        do {
            conversations = try await FishersAPI.conversations()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func open(_ conversation: ConversationSummary) async {
        openConversationId = conversation.id
        messages = []
        proposals = []
        agentSummary = nil
        await refreshThread()
        // Clearing the badge is best-effort; a failure here shouldn't surface.
        try? await FishersAPI.markRead(conversationId: conversation.id)
        await loadConversations()
    }

    func refreshThread() async {
        guard let id = openConversationId else { return }
        isLoading = true
        defer { isLoading = false }
        do {
            // The API returns newest first; the thread reads oldest at the top.
            messages = try await FishersAPI.messages(conversationId: id).reversed()
            proposals = try await FishersAPI.proposals(conversationId: id)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func send(_ body: String) async {
        guard let id = openConversationId else { return }
        let trimmed = body.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        isSending = true
        defer { isSending = false }
        do {
            let message = try await FishersAPI.postMessage(conversationId: id, body: trimmed)
            messages.append(message)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    /// Captain-only: have the assistant read the thread.
    func askAssistant() async {
        guard let id = openConversationId else { return }
        isThinking = true
        errorMessage = nil
        defer { isThinking = false }
        do {
            let analysis = try await FishersAPI.analyseConversation(id)
            agentSummary = analysis.summary
            proposals = analysis.proposals + proposals.filter { !$0.isPending }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func apply(_ proposal: AgentProposal) async {
        await decide(proposal) { try await FishersAPI.applyProposal($0) }
        await refreshThread()
    }

    func dismiss(_ proposal: AgentProposal) async {
        await decide(proposal) { try await FishersAPI.dismissProposal($0) }
    }

    private func decide(
        _ proposal: AgentProposal,
        _ work: (UUID) async throws -> AgentProposal
    ) async {
        do {
            let updated = try await work(proposal.id)
            if let index = proposals.firstIndex(where: { $0.id == updated.id }) {
                proposals[index] = updated
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func createConversation(title: String, clubId: UUID?) async {
        do {
            _ = try await FishersAPI.createConversation(title: title, clubId: clubId)
            await loadConversations()
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: flutter/lib/stores/chat_store.dart
import 'package:flutter/foundation.dart';

import '../models/fishers_models.dart';
import '../services/fishers_api.dart';
import '../services/network_service.dart';

/// Chat state — a port of `ios/Fishers/ViewModels/ChatStore.swift`.
///
/// The thread list, the open thread's messages, and the assistant's proposals.
/// The proposals are read but not yet acted on: applying one is a captain's
/// decision and belongs with the rest of the assistant work, which has not
/// been ported.
class ChatStore extends ChangeNotifier {
  List<ConversationSummary> _conversations = <ConversationSummary>[];
  List<ChatMessage> _messages = <ChatMessage>[];
  bool _isLoading = false;
  bool _isSending = false;
  String? _errorMessage;
  String? _openConversationId;

  List<ConversationSummary> get conversations =>
      List<ConversationSummary>.unmodifiable(_conversations);

  /// Oldest at the top, as a thread reads. The API pages newest-first — "the
  /// latest 50" is a LIMIT on a descending query — so what arrives is upside
  /// down, and iOS reverses it for the same reason.
  List<ChatMessage> get messages => List<ChatMessage>.unmodifiable(_messages);

  bool get isLoading => _isLoading;
  bool get isSending => _isSending;
  String? get errorMessage => _errorMessage;
  String? get openConversationId => _openConversationId;

  Future<void> loadConversations() async {
    _isLoading = true;
    notifyListeners();
    try {
      _conversations = await FishersAPI.conversations();
      _errorMessage = null;
    } on ApiException catch (error) {
      _errorMessage = error.friendlyMessage;
    } on Object catch (error) {
      _errorMessage = '$error';
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<void> open(ConversationSummary conversation) async {
    _openConversationId = conversation.id;
    _messages = <ChatMessage>[];
    notifyListeners();
    await refreshThread();
    // Clearing the badge is best-effort; a failure here should not surface.
    try {
      await FishersAPI.markRead(conversationId: conversation.id);
    } on Object {
      // The count corrects itself on the next load.
    }
    await loadConversations();
  }

  Future<void> refreshThread() async {
    final String? id = _openConversationId;
    if (id == null) return;
    _isLoading = true;
    notifyListeners();
    try {
      final List<ChatMessage> page = await FishersAPI.messages(conversationId: id);
      _messages = page.reversed.toList(growable: false);
      _errorMessage = null;
    } on ApiException catch (error) {
      _errorMessage = error.friendlyMessage;
    } on Object catch (error) {
      _errorMessage = '$error';
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  /// Somebody else wrote in the open thread: show it, and it has been read.
  Future<void> threadChanged() async {
    final String? id = _openConversationId;
    if (id == null) return;
    await refreshThread();
    try {
      await FishersAPI.markRead(conversationId: id);
    } on Object {
      // As above.
    }
  }

  Future<void> send(String body) async {
    final String? id = _openConversationId;
    final String trimmed = body.trim();
    if (id == null || trimmed.isEmpty) return;
    _isSending = true;
    notifyListeners();
    try {
      final ChatMessage message = await FishersAPI.postMessage(conversationId: id, body: trimmed);
      _messages = <ChatMessage>[..._messages, message];
      _errorMessage = null;
    } on ApiException catch (error) {
      _errorMessage = error.friendlyMessage;
    } on Object catch (error) {
      _errorMessage = '$error';
    } finally {
      _isSending = false;
      notifyListeners();
    }
  }

  void close() {
    _openConversationId = null;
    _messages = <ChatMessage>[];
  }
}
